hbase-server/src/main/java/org/apache/hadoop/hbase/master/procedure/AbstractSnapshottingStateMachineTableProcedure.java

```diff
@@ -0,0 +1,306 @@
+    boolean isDone;
+    try {
+      // isSnapshotDone will return true if the snapshot is completed and ready, false otherwise.
+      // It will also throw an HBaseSnapshotException if the snapshot failed, which is what we
+      // want, so this failure can propagate and fail the procedure execution.
+      isDone = snapshotManager.isSnapshotDone(snapshot);
+      LOG.debug("isSnapshotDone returns {}", isDone);
+    } catch (UnknownSnapshotException e) {
+      // We may not have requested the snapshot yet.
+      LOG.debug("isSnapshotDone throws UnknownSnapshotException, we have not requested it yet");
+      isDone = false;
+    } catch (IOException e) {
+      LOG.debug("isSnapshotDone throws an exception indicating a snapshotting failure", e);
+      throw e; // rethrow
+    }
+    if (!isDone) {
+      // A snapshot may be running on the table. It is either ours or someone else's but either way
+      // we can handle it using the same logic.
+      if (snapshotManager.isTakingSnapshot(getTableName())) {
+        LOG.debug("Snapshot is in progress for table {}", getTableName());
+      } else {
+        // Our snapshot is not in completed state and no snapshot is running on the table. This
+        // means we haven't made the request yet and need to do that now.
+        LOG.info("Taking recovery snapshot {} on {}", snapshotName, getTableName());
+        // We bypass the main entry point for SnapshotManager's take snapshot function in order
+        // to supply a custom snapshot handler that avoids taking an exclusive lock on the table.
+        // We already are holding the exclusive lock so that wouldn't work. So now we must upcall
+        // to any coprocessors registered on the snapshot hook to preserve that semantic. This
+        // should be refactored.
+        // prepareWorkingDirectory prepares the working directory for the snapshot. We must call
+        // this first.
+        snapshotManager.prepareWorkingDirectory(snapshot);
+        // call pre coproc hook
+        TableDescriptor desc = null;
+        MasterCoprocessorHost cpHost = env.getMasterCoprocessorHost();
+        MasterServices masterServices = env.getMasterServices();
+        org.apache.hadoop.hbase.client.SnapshotDescription snapshotPOJO = null;
+        if (cpHost != null) {
+          snapshotPOJO = ProtobufUtil.createSnapshotDesc(snapshot);
+          cpHost.preSnapshot(snapshotPOJO, desc, null);
+        }
+        if (snapshotHandler == null) {
+          TableState state = masterServices.getTableStateManager().getTableState(getTableName());
+          if (state.isEnabled()) {
+            snapshotHandler =
+              new NoLockEnabledTableSnapshotHandler(snapshot, masterServices, snapshotManager);
+          } else {
+            snapshotHandler =
+              new NoLockDisabledTableSnapshotHandler(snapshot, masterServices, snapshotManager);
+          }
+        }
+        snapshotManager.snapshotTable(snapshot, snapshotHandler);
```

Comment:
We will block here until the snapshot is done? I think we'd better schedule a sub procedure so we will release the procedure worker thread.

Reply:
No we do not block. We throw a ProcedureSuspendException and get the procedure rescheduled. I specifically designed this to not do the kind of blocking you are concerned about.

Reply:
So why not just submit a child procedure for doing the snapshoting? In this way, after the sub procedure finishes, the parent procedure will continue again, so we do not need to add these timeout/check logic here.